rapport.js

```diff
@@ -0,0 +1,239 @@
+'use strict';
+
+/**
+ * @typedef {object} Rapport
+ * @property {function} create
+ * @property {function} wrap
+ */
+
+/**
+ * @typedef {object} RapportWrapper
+ * @property {function} onOpen
+ * @property {function} onError
+ * @property {function} onClose
+ * @property {function} onMessage
+ * @property {function} close
+ * @property {function} send
+ * @property {function} request
+ */
+
+(function() {
+
+    /**
+     * Creates a Rapport instance.
+     *
+     * @param wsImplementation The websocket implementation to wrap.
+     * @param {object} [options] Optional options object.
+     * @param {function} [options.stringify] The function to use to convert messages to strings. Defaults to JSON.stringify.
+     * @param {function} [options.parse] The function to use to parse messages from strings. Defaults to JSON.parse.
+     * @param {function} [options.Promise] The promise implementation to use. Defaults to the ES6 Promise if found.
+     * @param {function} [options.generateRequestId] The function to use to generate request IDs. Defaults to a UUID.
+     * @return {Rapport} Rapport library.
+     */
+    function Rapport(wsImplementation, options) {
+        const opts = {
+            stringify: options.stringify || JSON.stringify,
+            parse: options.parse || JSON.parse,
+            Promise: options.Promise || (typeof Promise !== 'undefined') ? Promise : null,
```

Comment:
I don't think you're handling the case when promise is null properly. Unless you're relying on the user to specify callback in this case, but there seems to be nothing enforcing that.